### PR TITLE
Implement multi-store proof solidity library 

### LINKED ISSUE #1403: Implement multi-store proof solidity library
- [ ] Define library struct + `getAppHash` function
- [ ] Change input of `relayOracleState` to use library data
- [ ] Add testing

### PR DESCRIPTION
**Pull request by Benzbeeb**
Fixed #1403

### CODE REVIEW

File: bridges/evm/contracts/MultiStore.sol [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+//                         ________________[AppHash]_______________
+//                        /                                        \
+//             _______[I12]______                         ________[I13]________
+//            /                  \                       /                     \
+//       __[I8]__             __[I9]__              __[I10]__               __[I11]__
+//      /         \          /         \           /         \            /           \
+//    [I1]       [I2]     [I3]        [I4]       [I5]        [I6]       [I7]          [E]
+//   /   \      /   \    /    \      /    \     /    \       /   \     /    \
+// [0]   [1]  [2]   [3] [4]   [5]  [6]    [7] [8]    [9]   [A]   [B] [C]    [D]
+// [0] - acc     [1] - bank      [2] - capbility  [3] - distribution  [4] - evidence
```
Comment (taobun) [inline bridges/evm/contracts/MultiStore.sol]: capability

File: bridges/evm/contracts/MultiStore.sol [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+        return
+            Utils.merkleInnerHash( // [AppHash]
+                _self.accToMainStoresMerkleHash, // [I12]
+                Utils.merkleInnerHash( // [I13]
+                    Utils.merkleInnerHash( // [I10]
+                        Utils.merkleInnerHash( // [I5]
+                            _self.mintStoresMerkleHash, // [8]
+                            Utils.merkleLeafHash( // [9]
+                                abi.encodePacked(
+                                    hex"066f7261636c6520", // oracle prefix
```
Comment (taobun) [inline bridges/evm/contracts/MultiStore.sol]: // oracle prefix (uint8(6) + "oracle" + uint8(32))

**Review by taobun — CHANGES_REQUESTED** [review]

File: bridges/evm/contracts/MultiStore.sol [hunk 3]
```diff
@@ -0,0 +1,59 @@
+pragma solidity 0.5.14;
+import {Utils} from "./Utils.sol";
+
+
+// Computes Tendermint's application state hash at this given block. AppHash is actually a
+// Merkle hash on muliple stores. Luckily, we only care about "oracle" tree and all other
```
Comment (sorawit) [inline bridges/evm/contracts/MultiStore.sol]: This is not true anymore. We are not that lucky. Please read comment before copy-pasting

**Review by sorawit — CHANGES_REQUESTED** [review]

**Review by sorawit — DISMISSED** [review]

File: bridges/evm/test/Bridge.test.js [hunk 4]
```diff
@@ -5,233 +5,225 @@ const ReceiverMock = artifacts.require("ReceiverMock");
...
-    //         "0x12240A203A88B5657ABEF8D8CAB1AD0FB84C4796B1106E68F60B20EAFDE282B92E5270AF10012A0C0884F4E3F20510CCE8E1E801320962616E64636861696E", // _signedDataSuffix
-    //       ],
-    //     ]
-    //   );
-    // });
+    it("should accept correct state relay (All signatures)", async () => {
+      await this.bridge.relayOracleState(
+        "63407", // _blockHeight,
+        [
+          "0x2B14E5AB19A88F23B3F6C694659374DF21E21935A7C0B8565C4E7B43AAB3BB43", // accToMainStoresMerkleHash
```
Comment (taobun) [inline bridges/evm/test/Bridge.test.js]: change all variable name

File: bridges/evm/test/Bridge.test.js [hunk 5]
```diff
@@ -5,233 +5,225 @@ const ReceiverMock = artifacts.require("ReceiverMock");
 require("chai").should();
 
 contract("Bridge", ([_, owner, alice, bob]) => {
-  context("Checking oracle state relay (4 validators)", () => {
+  context("Checking oracle state relay (3 validators)", () => {
     beforeEach(async () => {
       this.bridge = await Bridge.new(
         [
-          ["0x652D89a66Eb4eA55366c45b1f9ACfc8e2179E1c5", 100],
```
Comment (taobun) [inline bridges/evm/test/Bridge.test.js]: Why remove this validator?

File: bridges/evm/test/Bridge.test.js [hunk 6]
```diff
@@ -5,233 +5,225 @@ const ReceiverMock = artifacts.require("ReceiverMock");
...
+          [
+            "0x32FA694879095840619F5E49380612BD296FF7E950EAFB66FF654D99CA70869E", // versionAndChainIdHash
+            "0x9F667F22196CFEBD590E3B178148942EF40C5B693336BF20608D5F21AA679ABC", // timeHash
+            "0xF665C04557D07EAACFC46F2E5B5D267C3B2C4EEAF395784F30CC84266816A89D", // lastBlockIDAndOther
+            "0xE3BB39F3B621939B734AF396D7CF5CCFDEDD4A3D16F139FD481874BC6F6E52B4", // nextValidatorHashAndConsensusHash
+            "0x6E340B9CFFB37A989CA544E6BB780A2C78901D3FB33738768511A30617AFA01D", // lastResultsHash
+            "0xD991DA4D4E69473CC75A4B819F9E07D4956671A6F4A74DF4CC16596FCBE68137" // evidenceAndProposerHash
+          ],
+          "0x6E080211AFF700000000000022480A20", // _signedDataPrefix
+          [
```
Comment (taobun) [inline bridges/evm/test/Bridge.test.js:167]: Invalid signature not missing signatures

File: bridges/evm/test/Bridge.test.js [hunk 7]
```diff
@@ -5,233 +5,225 @@ const ReceiverMock = artifacts.require("ReceiverMock");
...
-    //           "0x12240A203A88B5657ABEF8D8CAB1AD0FB84C4796B1106E68F60B20EAFDE282B92E5270AF10012A0C0884F4E3F20510CCE8E1E801320962616E64636861696E", // _signedDataSuffix
-    //         ],
-    //       ]
-    //     ),
-    //     "INSUFFICIENT_VALIDATOR_SIGNATURES"
-    //   );
-    // });
+    it("should revert if sum of validator powers is less than 2/3)", async () => {
+      await this.bridge.updateValidatorPowers(
+        [["0x88e1cd00710495EEB93D4f522d16bC8B87Cb00FE", 300]],
```
Comment (taobun) [inline bridges/evm/test/Bridge.test.js]: This address is a missing signature?

**Review by taobun — CHANGES_REQUESTED** [review]

File: bridges/evm/test/Bridge.test.js (hunk 7 above)
Comment (Benzbeeb) [reply bridges/evm/test/Bridge.test.js]: ya

File: bridges/evm/test/Bridge.test.js [hunk 8]
```diff
@@ -17,221 +17,183 @@ contract("Bridge", ([_, owner, alice, bob]) => {
...
-    // });
+    it("should accept correct state relay (validator power is more than 2/3)", async () => {
+      await this.bridge.relayOracleState(
+        "63407", // _blockHeight,
+        [
+          "0x2B14E5AB19A88F23B3F6C694659374DF21E21935A7C0B8565C4E7B43AAB3BB43", // accToMemCapStoresMerkleHash
+          "0x06223635C7D081E76F30998DA7D601AB9AB8602C9AE7095E319C68F4CE2F5F73", // mintStoresMerkleHash
+          "0xC744B497E2569517AA2F053E2A4E09C67427DC0FD8840048C607AF8F17D9100C", // oracleIAVLStateHash
+          "0xAE658B0EA9B1C87323D0FA810ADBE9DB91BE0BAEA62D29F376488666ECA9BBEB", // paramsAndSlashingStoresMerkleHash
+          "0xAA66B682A3FE96C372A095A43F2E73B33E89D835DB3BBF2496546328A7BD4961" // stakingToUpgradeStoresMerkleHash
```
Comment (taobun) [inline bridges/evm/test/Bridge.test.js]: StakingAndUpgradeStoresMerkleHash

**Review by taobun — CHANGES_REQUESTED** [review]

File: bridges/evm/contracts/MultiStore.sol [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+// root hash, since we only want to validate the correctness of [9] In fact, only
+// [I11], [8], [9], [I6], and [I10] are needed in order to compute [AppHash].
+
+library MultiStore {
+    struct Data {
+        bytes32 accToMemCapStoresMerkleHash; // [I11]
+        bytes32 mintStoresMerkleHash; // [8]
+        bytes32 oracleIAVLStateHash; // [9]
+        bytes32 paramsAndSlashingStoresMerkleHash; // [I6]
+        bytes32 StakingAndUpgradeStoresMerkleHash; // [I10]
```
Comment (taobun) [inline bridges/evm/contracts/MultiStore.sol]: camel Pls

File: bridges/evm/test/Bridge.test.js [hunk 10]
```diff
@@ -17,221 +17,177 @@ contract("Bridge", ([_, owner, alice, bob]) => {
...
-    // });
+    it("should accept correct state relay (validator power is more than 2/3)", async () => {
+      await this.bridge.relayOracleState(
+        "63407", // _blockHeight,
+        [
+          "0x2B14E5AB19A88F23B3F6C694659374DF21E21935A7C0B8565C4E7B43AAB3BB43", // accToMemCapStoresMerkleHash
+          "0x06223635C7D081E76F30998DA7D601AB9AB8602C9AE7095E319C68F4CE2F5F73", // mintStoresMerkleHash
+          "0xC744B497E2569517AA2F053E2A4E09C67427DC0FD8840048C607AF8F17D9100C", // oracleIAVLStateHash
+          "0xAE658B0EA9B1C87323D0FA810ADBE9DB91BE0BAEA62D29F376488666ECA9BBEB", // paramsAndSlashingStoresMerkleHash
+          "0xAA66B682A3FE96C372A095A43F2E73B33E89D835DB3BBF2496546328A7BD4961" // StakingAndUpgradeStoresMerkleHash
```
Comment (taobun) [inline bridges/evm/test/Bridge.test.js]: ditto

File: bridges/evm/test/Bridge.test.js [hunk 11]
```diff
@@ -17,221 +17,177 @@ contract("Bridge", ([_, owner, alice, bob]) => {
...
+            "0x9F667F22196CFEBD590E3B178148942EF40C5B693336BF20608D5F21AA679ABC", // timeHash
+            "0xF665C04557D07EAACFC46F2E5B5D267C3B2C4EEAF395784F30CC84266816A89D", // lastBlockIDAndOther
+            "0xE3BB39F3B621939B734AF396D7CF5CCFDEDD4A3D16F139FD481874BC6F6E52B4", // nextValidatorHashAndConsensusHash
+            "0x6E340B9CFFB37A989CA544E6BB780A2C78901D3FB33738768511A30617AFA01D", // lastResultsHash
+            "0xD991DA4D4E69473CC75A4B819F9E07D4956671A6F4A74DF4CC16596FCBE68137" // evidenceAndProposerHash
+          ],
+          "0x6E080211AFF700000000000022480A20", // _signedDataPrefix
+          [
+            [
+              "0x62A900783D67A397A0CE36C82108122B97ACF88034CA917209937E2FF29EDB1C", // r
```
Comment (taobun) [inline bridges/evm/test/Bridge.test.js:169]: Please comment on what signature is wrong

**Review by taobun — CHANGES_REQUESTED** [review]

**Review by taobun — DISMISSED** [review]

**Review by taobun — APPROVED** [review]

**Review by sorawit — APPROVED** [review]

**Pull request merged**